core/trino-main/src/test/java/io/trino/server/remotetask/TestHttpRemoteTask.java

```diff
@@ -369,6 +378,47 @@ public void testOutboundDynamicFilters()
         httpRemoteTaskFactory.stop();
     }
 
+    @Test(timeOut = 300000)
+    public void testAdaptiveRemoteTaskRequestSize()
+            throws Exception
+    {
+        AtomicLong lastActivityNanos = new AtomicLong(System.nanoTime());
+        TestingTaskResource testingTaskResource = new TestingTaskResource(lastActivityNanos, FailureScenario.NO_FAILURE);
+
+        Session session = testSessionBuilder()
+                .setCatalog("tpch")
+                .setSchema(TINY_SCHEMA_NAME)
+                .setSystemProperty(ENABLE_ADAPTIVE_REMOTE_TASK_REQUEST_SIZE, "true")
+                .setSystemProperty(MAX_REMOTE_TASK_REQUEST_SIZE, "8kB")
+                .setSystemProperty(REMOTE_TASK_REQUEST_SIZE_HEADROOM, "2kB")
+                .setSystemProperty(REMOTE_TASK_GUARANTEED_SPLITS_PER_REQUEST, "2")
+                .build();
+        HttpRemoteTaskFactory httpRemoteTaskFactory = createHttpRemoteTaskFactory(testingTaskResource);
+
+        RemoteTask remoteTask = createRemoteTask(httpRemoteTaskFactory, ImmutableSet.of(), session);
+
+        testingTaskResource.setInitialTaskInfo(remoteTask.getTaskInfo());
+        remoteTask.start();
+
+        Multimap<PlanNodeId, Split> splits = HashMultimap.create();
+        for (int i = 0; i < 1000; i++) {
+            splits.put(TABLE_SCAN_NODE_ID, new Split(TEST_CATALOG_HANDLE, TestingSplit.createLocalSplit()));
+        }
+        remoteTask.addSplits(splits);
+
+        poll(() -> testingTaskResource.getTaskSplitAssignment(TABLE_SCAN_NODE_ID) != null);
```

Comment:
Even though `splits` is divided into smaller requests, it will eventually send all 1000 splits.

Let me add `poll(() -> testingTaskResource.getTaskSplitAssignment(TABLE_SCAN_NODE_ID).getSplits().size() == 1);` to check whether `splits` is actually divided or not.

Reply:
Can we test if request size did not exceed our maximum?